Downgrade "context canceled" errors to Info instead of Error

Mircoxi opened the pull request:
What: 

This PR simply downgrades the "context canceled" errors as part of piece uploads from Error to Info.

Why:

"context canceled" is a normal state of operation during uploads, and results in log spam, even when a user has elevated their default logging preference. Since it's ostensibly not an error, it shouldn't be marked as such in the logs. This would also close #5799, as well as making many people on the forums happy!

There's an argument to be made that this should be in the section above instead, since technically it's a cancellation rather than a failure, but I assumed that if it wasn't in that section already, it was for a reason. I'm happy to adjust that if preferred. 


Please describe the tests:

None added, since this is a simple logging level change.
 
Please describe the performance impact:

Almost zero. 

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

Mircoxi commented:
Signed CLA a few minutes prior to committing!

liori commented on storagenode/piecestore/endpoint.go:
```diff
@@ -358,7 +359,13 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
 			mon.IntVal("upload_failure_duration_ns").Observe(uploadDuration)
 			mon.FloatVal("upload_failure_rate_bytes_per_sec").Observe(uploadRate)
-			endpoint.log.Error("upload failed", zap.Stringer("Piece ID", limit.PieceId), zap.Stringer("Satellite ID", limit.SatelliteId), zap.Stringer("Action", limit.Action), zap.Error(err), zap.Int64("Size", uploadSize), remoteAddrLogField)
+			if errors.Is(err, context.Canceled) {
```
I also observe _context canceled_ for downloads. Do you want to also patch it?

Mircoxi replied:
Done, thanks for the heads up! It didn't occur to me to check for that. I notice the CI is failing right now - on my end, testing just that module shows a pass, and (if I'm understanding the output correctly - I may not be, I'm not super-experienced with Go just yet) the changes to this section are passing on Jenkins' end too, but failing as a whole. I'll some time aside later to debug this properly, though a more experienced pair of eyes is always welcome!

Mircoxi commented:
Checking the failing tests and the history available to me on the Jenkins log, it seems commit https://github.com/storj/storj/commit/03c52f184ece3b9de99279e0b706cac14dc3931f introduced the bug that's causing test failures in this PR (it seems to be a schema change, which is consistent with what the errors are showing - perhaps the tests haven't been updated to account for it yet?), but it does appear that the tests for the parts I changed are passing. I'd appreciate some guidance on how to proceed given the failing tests. 

Also, I resubmitted the CLA form since I changed my email address - but I'm confirming that I've signed it. I'll hold off on any more updating the code on my end to keep the bot happy.

thepaul commented on storagenode/piecestore/endpoint.go:
```diff
@@ -358,7 +359,13 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
-			endpoint.log.Error("upload failed", zap.Stringer("Piece ID", limit.PieceId), zap.Stringer("Satellite ID", limit.SatelliteId), zap.Stringer("Action", limit.Action), zap.Error(err), zap.Int64("Size", uploadSize), remoteAddrLogField)
+			if errors.Is(err, context.Canceled) {
+				// Context cancellation is common in normal operation, and shouldn't throw a full error.
+				endpoint.log.Info("upload failed", zap.Stringer("Piece ID", limit.PieceId), zap.Stringer("Satellite ID", limit.SatelliteId), zap.Stringer("Action", limit.Action), zap.Error(err), zap.Int64("Size", uploadSize), remoteAddrLogField)
```
I would suggest going even farther and taking out the `zap.Error(err)` parameter. You could add `"(context canceled)"` to the log message itself instead. Our errors tend to be very long-winded and the stack trace is unlikely to be important.

Maybe we could log the error and stack trace in a Debug call below as well, just in case it ever _is_ important.

Same recommendation for the Download case.

Mircoxi commented:
Good call on the shorter messages! I changed the originals to debug level, and then took the liberty of making the version logged at Info leave out everything but the piece ID and a short message telling the user that they lost the race, since I've seen a fair bit of confusion and worry that "context canceled" is a more serious error than it is.

thepaul commented on storagenode/piecestore/endpoint.go:
```diff
@@ -643,7 +645,8 @@ func (endpoint *Endpoint) Download(stream pb.DRPCPiecestore_DownloadStream) (err
 			if errors.Is(err, context.Canceled) {
-				endpoint.log.Info("download failed", zap.Stringer("Piece ID", limit.PieceId), zap.Stringer("Satellite ID", limit.SatelliteId), zap.Stringer("Action", limit.Action), zap.Int64("Offset", chunk.Offset), zap.Int64("Size", downloadSize), zap.String("Remote Address", remoteAddr), zap.Error(err))
+				endpoint.log.Info("download canceled (context canceled)", zap.Stringer("Piece ID", limit.PieceId), zap.String("Reason", "Someone else downloaded the required piece faster."))
+				endpoint.log.Debug("download canceled (context canceled)", zap.Stringer("Piece ID", limit.PieceId), zap.Stringer("Satellite ID", limit.SatelliteId), zap.Stringer("Action", limit.Action), zap.Int64("Offset", chunk.Offset), zap.Int64("Size", downloadSize), zap.String("Remote Address", remoteAddr), zap.Error(err))
```
We probably don't need the "` (context canceled)`" hint on this line, since the err is still being provided.

thepaul commented on storagenode/piecestore/endpoint.go:
```diff
@@ -361,7 +361,9 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
 			if errors.Is(err, context.Canceled) {
 				// Context cancellation is common in normal operation, and shouldn't throw a full error.
-				endpoint.log.Info("upload failed", zap.Stringer("Piece ID", limit.PieceId), zap.Stringer("Satellite ID", limit.SatelliteId), zap.Stringer("Action", limit.Action), zap.Error(err), zap.Int64("Size", uploadSize), remoteAddrLogField)
+				endpoint.log.Info("upload canceled (context canceled)", zap.Stringer("Piece ID", limit.PieceId), zap.String("Reason", "Someone else uploaded the required piece faster."))
```
It's also possible to get context.Canceled when the node is shutting down, so those log lines could be confusing when a bunch of them show up at the end of a log. And unfortunately it's not trivial to distinguish those cases. What do you think about this-

```go
endpoint.log.Info("upload canceled (race lost or node shutdown)", zap.Stringer("Piece ID", limit.PieceId))
```

Mircoxi replied:
My bad there! I didn't find any documentation about that possibility, I've gone ahead and corrected it.
